csrc/preseg_passes/propagate_shardings.cpp

```diff
@@ -301,27 +133,257 @@ std::unordered_set<ParallelType> getParallelTypesToPropagate(
+        ? std::make_pair(target_id, target)
+        : std::make_pair(ref_id, ref);
+
+    if (hasRootToLogicalTransform(consumer_id, consumer_tv)) {
+      validate_split(split, target_id);
+    }
+
+    auto it = transformed_loop.erase(target_id).second;
+    auto [outer, inner] =
+        IterDomain::split(target_id, split->factor(), split->innerSplit());
+
+    transformed_loop.insert(it, outer, std::monostate());
+    transformed_loop.insert(it, inner, std::monostate());
+
+    // Add mapping between ref and target for the propagated DID split.
+    // This is used to propagate 2D sharding and parallelization.
+    ref2target[split->outer()] = outer;
+    ref2target[split->inner()] = inner;
+  }
+
+  // Parallelize based on the ref2target map.
+  for (IterDomain* device_id : device_ids) {
+    NVF_ERROR(
+        ref2target.find(device_id) != ref2target.end(),
+        "Failed to propagate ",
+        device_id,
+        " to ",
+        target);
+    IterDomain* target_id = ref2target.at(device_id);
+    target_id->parallelize(device_id->getParallelType());
+  }
+
+  auto new_loop = std::views::keys(transformed_loop);
+  target->setLoopDomain({new_loop.begin(), new_loop.end()});
+}
+
+void propagateDIDTransform(
+    const TensorView* ref,
+    TensorView* tv,
+    const std::unordered_set<ParallelType>& selected_parallel_types,
+    PropagateDirection direction) {
+  tv->setDeviceMesh(ref->getDeviceMesh());
+
+  std::unordered_set<IterDomain*> device_ids;
+  const std::unordered_map<IterDomain*, IterDomain*> ref2target =
+      getRef2TargetMap(ref, tv, direction);
+
+  for (IterDomain* maybe_did : ref->getLoopDomain()) {
+    if (selected_parallel_types.count(maybe_did->getParallelType()) == 0) {
+      continue;
+    }
+    // Get input of maybe_did in the root / logical domain
+    // that will be present in ref2target mapping.
+    std::unordered_set<IterDomain*> inputs =
+        direction == PropagateDirection::kForward
+        ? getInputsInTargetDomain({maybe_did}, ref->getLogicalDomain())
+        : getInputsInTargetDomain({maybe_did}, ref->getMaybeRootDomain());
+    NVF_ERROR(
+        inputs.size() == 1,
+        "Expected one input for ",
+        maybe_did,
+        " in the root / logical domain. Got ",
+        inputs.size(),
+        " inputs.");
+    IterDomain* ref_id = *inputs.begin();
+    IterDomain* target_id = getOrDefault(ref2target, ref_id);
+    if (target_id == nullptr) {
+      continue;
+    }
+
+    // Get the outermost logical ID corresponding to the target ID.
+    // This is the ID to which the sharding will be propagated.
+    // For e.g., if ref [h] -> target [a, h/a], then the outermost logical ID
+    // is `a`.
+    IterDomain* outermost_target_id = getOutermostLogicalId(target_id, tv);
+    // Skip if the target is parallelized or not in the loop domain (i.e.
+    // further transformed).
+    if (outermost_target_id->isParallelized() ||
```

Comment:
```
if (A || B) {
  continue;
}
```

==> 

```
if (A) {
  continue;
}
if (B) {
  continue;
}
```
So readers can focus on one condition at a time. It's also easier to gdb which condition triggers the continue.